addons/mail/static/src/discuss/voice_message/common/voice_recorder.js

```diff
@@ -0,0 +1,133 @@
+    static template = "mail.VoiceRecorder";
+
+    microphone;
+    recorder;
+    startTimeStamp;
+
+    setup() {
+        this.state = useState({
+            isActionPending: false,
+            recording: false,
+            intervalId: false,
+            elapsed: "00 : 00",
+        });
+        this.notification = useService("notification");
+        this.userSettings = useService("mail.user_settings");
+        this.isSafari = /^((?!chrome|android).)*safari/i.test(navigator.userAgent);
+        onWillUnmount(() => {
+            this.unmount = true;
+            this.stopRecording();
+        });
+    }
+
+    filename() {
+        return (
+            "Voice-" +
+            new Date().toISOString().split("T")[0] +
+            "-" +
+            Math.floor(Math.random() * 100000) +
+            ".mp3"
+        );
+    }
+
+    async startRecording(ev) {
+        if (this.state.isActionPending) {
+            return;
+        }
+        this.state.isActionPending = true;
+        if (!this.microphone) {
+            try {
+                this.microphone = await browser.navigator.mediaDevices.getUserMedia({
+                    audio: this.userSettings.audioConstraints,
+                });
+            } catch {
+                this.notification.add(
+                    sprintf(_t('"%(hostname)s" needs to access your microphone'), {
+                        hostname: window.location.host,
+                    }),
+                    { type: "warning" }
+                );
+                this.state.isActionPending = false;
+                return;
+            }
+        }
+        this.state.elapsed = "00 : 00";
+        this.props.onchangeRecording();
+        this.state.recording = true;
+        let totalSeconds = 0;
+        this.state.intervalId = setInterval(() => {
+            ++totalSeconds;
+            const second = totalSeconds % 60;
+            const minute = Math.floor(totalSeconds / 60) % 60;
+            this.state.elapsed =
+                (minute < 10 ? "0" + minute : minute) +
+                " : " +
+                (second < 10 ? "0" + second : second);
+            if (totalSeconds == 60) {
+                this.notification.add(
+                    _t("The duration of voice messages is limited to 1 minute."),
+                    { type: "warning" }
+                );
+                this.stopRecording();
+            }
+        }, 1000);
+        const mimeType = this.isSafari ? "audio/mp4" : "audio/webm";
+        this.recorder = new MediaRecorder(this.microphone, { mimeType, audioBitsPerSecond: 7400 });
+        this.startTimeStamp = ev.timeStamp;
+        this.recorder.ondataavailable = (e) => {
+            if (this.unmount) {
+                return;
+            }
+            const blob = new Blob([e.data], {
+                type: this.recorder.mimeType || mimeType,
+            });
+            this.duration = e.timeStamp - this.startTimeStamp;
```

Comment:
`this.duration` and `this.startTimeStamp` can be limited to the scope of `startRecording()`